projects/wizard_of_wikipedia/generator/agents.py

```diff
@@ -0,0 +1,358 @@
+
+TOKEN_DIALOG = '__dialog__'
+
+
+# extend the default batch
+class Batch(TABatch):
+    def __init__(
+        self,
+        # needed for evaluation metrics
+        checked_sentence=None,
+        # needed for model
+        cs_ids=None,
+        know_vec=None,
+        ck_mask=None,
+        use_cs_ids=None,
+        knowledge=None,
+        **kwargs,
+    ):
+        super().__init__(
+            checked_sentence=checked_sentence,
+            cs_ids=cs_ids,
+            know_vec=know_vec,
+            ck_mask=ck_mask,
+            use_cs_ids=use_cs_ids,
+            knowledge=knowledge,
+            **kwargs,
+        )
+
+
+DEFAULT_OPTS = {
+    "learningrate": 3e-4,
+    "optimizer": "adam",
+    "lr_scheduler": "invsqrt",
+    "warmup_updates": 1000,
+    "clip_norm": 0.1,
+    "ffn_size": 512,
+    "embedding_size": 256,
+    "n_heads": 2,
+    "relu_dropout": 0.2,
+    "attention_dropout": 0.2,
+    "n_layers": 3,
+    "truncate": 128,
+    "dict_textfields": "text,labels,chosen_topic,checked_sentence,knowledge,title",
+}
+
+
+class _GenericWizardAgent(TransformerGeneratorAgent):
+    # @classmethod
+    # def dictionary_class(cls):
+    #     import parlai_internal.projects.wizard.generator.agents as a
+    #     return a._WizardGeneratorDictionary
+
+    # todo:
+    # - dict agent stuff
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        argparser.set_defaults(**DEFAULT_OPTS)
+        super(_GenericWizardAgent, cls).add_cmdline_args(argparser)
+
+    def batchify(self, obs_batch):
+        batch = super().batchify(obs_batch)
+        reordered_observations = [obs_batch[i] for i in batch.valid_indices]
+
+        checked_sentences = []
+        for obs in reordered_observations:
+            checked_sentence = '{} {} {}'.format(
+                obs.get('title', ''), TOKEN_KNOWLEDGE, obs.get('checked_sentence', '')
+            )
+            checked_sentences.append(checked_sentence)
+
+        batch['checked_sentence'] = checked_sentences
+        return batch
+
+
+class TwoStageAgent(_GenericWizardAgent):
+    def observe(self, obs):
+        if 'text' not in obs:
+            return obs
+
+        # TODO: resolve this with #1421
+        # get the dialog stuff
+        reply = self.last_reply()
+        self.observation = self.get_dialog_history(obs, reply=reply)
+        # we need to store the old text so that we can restore it
+        oldtext = obs['text']
+
+        # now we want to force prepend the knowledge stuff
+        fields = []
+        if 'chosen_topic' in obs:
+            fields += [obs['title']]
+        if 'checked_sentence' in obs:
+            fields += [TOKEN_KNOWLEDGE, obs['checked_sentence']]
+        if obs['text'] != '':
+            fields += [TOKEN_DIALOG, obs['text']]
+        obs['text'] = ' '.join(fields)
+
+        # now vectorize with the extra knowledge. It'll all get stored in the
+        # text_vec operation, etc
+        self.vectorize(
```

Comment:
i guess it may not be *necessary* but is it worth setting `obs = self.vectorize(...)` rather than just calling it here? it's not 100% intuitive that vectorize modifies `obs`